### PR TITLE
[ADD] mass_mailing: add a design tab to apply global styles to mailing

### PR DESCRIPTION
**Pull request by Zynton**
This adds a design tab to mass_mailing, allowing the user to apply global styles to their mailing such as colors and fonts to headings for instance. This tab replaces the "select a template" tab.

task-2789287

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mass_mailing/static/src/js/mass_mailing_snippets.js [hunk 1]
```diff
@@ -151,4 +242,215 @@ options.registry.ImageTools.include({
...
+     * @param {string} [params.toggle] if 'true', will remove the property if
+     *                                 its value is already the one it's being
+     *                                 set to
+     * @param {string} [params.activeValue] the value to set, if `widgetValue`
+     *                                      is not defined.
+     * @returns {Promise|undefined}
+     */
+    customizeCssProperty(previewMode, widgetValue, params) {
+        if (!params.selectorText || !params.cssProperty) {
+            return;
```
Comment (dmo-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_snippets.js:326]: I'm curious, when can this happen in practice ?

File: addons/mass_mailing/static/src/js/mass_mailing_snippets.js [hunk 2]
```diff
@@ -151,4 +242,215 @@ options.registry.ImageTools.include({
...
+                // The rule doesn't exist: create it.
+                this.styleSheet.insertRule(`${selector} {
+                    ${params.cssProperty}: ${value} !important;
+                }`);
+            }
+        }
+        this._commitCss();
+    },
+    /**
+     * Option method to set a css property in the mailing's custom stylesheet.
```
Comment (dmo-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_snippets.js:360]: Looks like a copy paste of the previous method documentation.

File: addons/mass_mailing/views/assets.xml [hunk 3]
```diff
@@ -23,7 +23,7 @@
                 box-sizing: border-box !important;
             }
             .o_layout :not(.fa) {
-                font-family: Arial, sans-serif !important;
+                font-family: Arial,Helvetica Neue,Helvetica,sans-serif;
```
Comment (dmo-odoo) [inline addons/mass_mailing/views/assets.xml:26]: Is this line really relevant to this particular commit ? I don't get it.

File: addons/mass_mailing/views/snippets_themes.xml [hunk 4]
```diff
@@ -537,6 +522,277 @@
...
+                            data-selector-text=".s_hr > hr"
+                            data-no-transparency="true"
+                            data-color-prefix="border-"/>
+        </we-row>
+        <we-select string="⌙ Width" data-selector-text=".s_hr > hr" data-css-property="width">
+            <we-button data-customize-css-property="25%">25%</we-button>
+            <we-button data-customize-css-property="50%">50%</we-button>
+            <we-button data-customize-css-property="75%">75%</we-button>
+            <we-button data-customize-css-property="100%">100%</we-button>
+        </we-select>
```
Comment (dmo-odoo) [inline addons/mass_mailing/views/snippets_themes.xml:795]: I skipped over these lines as I suspect they will probably be changed during a UX redesign anyway ... :p

File: addons/web_editor/static/src/js/backend/convert_inline.js [hunk 5]
```diff
@@ -336,10 +336,6 @@ function classToStyle($editable, cssRules) {
                 writes.push(() => { node.style[styleName] = ''; });
             }
         }
-        // Ignore font-family (mail-safe font declared in <head>)
-        if ('font-family' in css) {
-            delete css['font-family'];
-        }
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/backend/convert_inline.js:342]: So we *do* accept not mail-safe fonts now, right ?

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 6]
```diff
@@ -3087,13 +3087,17 @@ const SnippetOptionWidget = Widget.extend({
      * @returns {Promise|undefined}
      */
     selectClass: function (previewMode, widgetValue, params) {
+        const $target = params.target ? this.options.wysiwyg.getEditable().find(params.target) : this.$target;
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3090]: It is not clear to me why this change is suddenly needed new. Same with the other methods below which have the same change. Did you *add* the concept of `params.target` or was it existing before but not properly supported in these particular functions ?

File: addons/mass_mailing/static/src/js/mass_mailing_snippets.js [hunk 7]
```diff
@@ -1,10 +1,101 @@
...
 const {ColorpickerWidget} = require('web.Colorpicker');
+const SelectUserValueWidget = options.userValueWidgetsRegistry['we-select'];
+const weUtils = require('web_editor.utils');
 const {_t} = require('web.core');
 
+//--------------------------------------------------------------------------
+// Constants
+//--------------------------------------------------------------------------
+
+const _getFontName = fontFamily => fontFamily.split(',')[0].replace(/"/g, '').replace(/([a-z])([A-Z])/g, (v, a, b) => `${a} ${b}`).trim();
```
[the remark this answers is not in the document]
Comment (Zinston) [reply addons/mass_mailing/static/src/js/mass_mailing_snippets.js:14]: Well that's true. I like it like this since it's declared at the top of the file and used further down. That way we can't confuse it for an util that might have been imported.

File: addons/mass_mailing/static/src/js/mass_mailing_snippets.js (hunk 1 above)
Comment (Zinston) [reply addons/mass_mailing/static/src/js/mass_mailing_snippets.js:336]: IIRC, when resetting after a preview.

File: addons/mass_mailing/static/src/js/mass_mailing_snippets.js (hunk 2 above)
Comment (Zinston) [reply addons/mass_mailing/static/src/js/mass_mailing_snippets.js]: Oops

File: addons/mass_mailing/views/assets.xml (hunk 3 above)
Comment (Zinston) [reply addons/mass_mailing/views/assets.xml:26]: We add the option to change the font family and this is now the default font family so I changed it in the general css as well.

File: addons/mass_mailing/views/snippets_themes.xml (hunk 4 above)
Comment (Zinston) [reply addons/mass_mailing/views/snippets_themes.xml:795]: :eyes:

File: addons/web_editor/static/src/js/backend/convert_inline.js (hunk 5 above)
Comment (Zinston) [reply addons/web_editor/static/src/js/backend/convert_inline.js:342]: No but we can now change to font to another mail-safe font.

File: addons/mass_mailing/static/src/js/mass_mailing_snippets.js [hunk 8]
```diff
@@ -329,28 +339,37 @@ options.registry.DesignTab = options.Class.extend({
...
         }
-        for (const selector of this._getSelectors(params.selectorText)) {
+        const selectors = this._getSelectors(params.selectorText);
+        if (params.cssProperty === 'font-family') {
+            // Ensure font-family gets passed to all descendants.
+            selectors.push(...selectors.map(selector => selector + ' *'));
+        }
+        const firstSelector = selectors[0].replace(CSS_PREFIX, '').trim();
+        for (const selector of selectors) {
+            const important = PRIORITY_STYLES[firstSelector].includes(params.cssProperty) ? ' !important' : '';
```
Comment (dmo-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_snippets.js:349]: I'm torn. On one hand I'd call it `priority` rather than `important` because sometimes `important` is not `!important`. But on the other hand, it makes it look like there might be more to it while really there is just `!important` or nothing... What do you think ?

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 6 above)
Comment (Zinston) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3090]: `this.$target` is defined for the whole snippet option. Here the option is the whole design tab so I needed a way to target a specific element from within the button. This makes sense to me since you might want an option that has a couple of buttons which each target a different element. Why not...?

File: addons/mass_mailing/static/src/js/mass_mailing_snippets.js (hunk 8 above)
Comment (Zinston) [reply addons/mass_mailing/static/src/js/mass_mailing_snippets.js:349]: I'm ok with either honestly. `getPropertyPriority` returns `'important` or `''` so to be coherent with the DOM API I _could_ indeed use `priority` if you prefer.

Comment (dmo-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_snippets.js:349]: Very good point ! Let's go with `priority` then. 👍

**Comment by dmo-odoo** [comment]
@robodoo rebase-ff r+

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 9]
```diff
@@ -3537,6 +3545,10 @@ const SnippetOptionWidget = Widget.extend({
      * @returns {Promise<string|undefined>|string|undefined}
      */
     _computeWidgetState: async function (methodName, params) {
+        const $target = params.target ? this.options.wysiwyg.getEditable().find(params.target) : this.$target;
+        if (!$target[0]) {
+            return;
+        }
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3551]: @Zinston You copy/paste that code in multiple generic methods, meaning their override would have to do the same, meaning that apparently-generic feature is only available to a bunch of generic methods, meaning you have to modify `this.$target` into `$target` in the whole implementation (breaking history)... That also complexifies the implementation of the method with a new "do I have the right `$target`?" at the beginning, things you should not have to care about when implementing an option method behavior (`selectStyle` = you want to force some style on the `$target`... that's it, not first asking yourself if you have the right `$target`).
Don't you think that option should have been coded on the caller side? That's what we do for the `apply-to` parameter... which is a near-duplicate of your new `target` parameter (one search in the whole DOM, one search in the original `$target`).

The minimum for me would to code it the same way as `apply-to` with a similar name. But what I think should be done is using the `apply-to` parameter as it is currently implemented and review the `this.$target` of your specific mass mailing option to be the whole editable area (maybe just a call to `setTarget` in the `init` would be enough?). Your new semi-generic parameter encourages a bad practice of having widgets displayed for a section A but changing a section B which could be anywhere in the page, far away from A => that's not the case for your specific option, the more the reason to code that specifically for your option by changing the base `$target` of your "Design" option as I suggest.

If you agree, could you make this a priority? I have small conflicts because of this, would be easier you change this first (otherwise I will be the one sending you conflicts by merging my resolved ones ^^).

Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3551]: @Zinston You copy/paste that code in multiple generic methods, meaning their override would have to do the same, meaning that apparently-generic feature is only available to a bunch of generic methods, meaning you have to modify `this.$target` into `$target` in the whole implementation (breaking history)... That also complexifies the implementation of the method with a new "do I have the right `$target`?" at the beginning, things you should not have to care about when implementing an option method behavior (`selectStyle` = you want to force some style on the `$target`... that's it, not first asking yourself if you have the right `$target`).
Don't you think that option should have been coded on the caller side? That's what we do for the `apply-to` parameter... which is a near-duplicate of your new `target` parameter (one search in the whole DOM, one search in the original `$target`).

The minimum for me would to code it the same way as `apply-to` with a similar name. But what I think should be done is using the `apply-to` parameter as it is currently implemented and review the `this.$target` of your specific mass mailing option to be the whole editable area (maybe just a call to `setTarget` in the `init` would be enough?). Your new semi-generic parameter encourages a bad practice of having widgets displayed for a section A but changing a section B which could be anywhere in the page, far away from A => that's not the case for your specific option, the more the reason to code that specifically for your option by changing the base `$target` of your "Design" option as I suggest.

If you agree, could you make this a priority? I have small conflicts because of this, would be easier you change this first (otherwise I will be the one sending you conflicts by merging my resolved ones ^^).

cc @dmo-odoo

Comment (Zinston) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3551]: @qsm-odoo I think you're right. We had discussed the fact that we probably would have to do something about this in a separate task (David left on vacation today and the freeze is next week so we had to leave that for later in order for the task to be ready on time). I'll start working on it asap so as to minimize issues for you. Sorry for the inconvenience.

Comment (Zinston) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3551]: @qsm-odoo I'm a bit confused with this `apply-to` thingy. To compute the visibility of the option, in several places we do `this.$(params.applyTo).eq(0);` but in that context `params.applyTo` is a selector for an element in the editable area whereas `this.$el` is the option in the sidebar. So it never finds it. I tried to find a case in Website where it does and actually never got past the "if not then return". Is it broken or am I missing something?

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3551]: > in several places we do [...]

We use it at 3 places:
- The caller of option methods (e.g. `selectStyle`)
- The caller of `_computeWidgetState`
- The caller of `_computeWidgetVisibility`

> this.$(params.applyTo).eq(0); but in that context params.applyTo is a selector for an element in the editable area whereas this.$el is the option in the sidebar.

I think you misses the fact that `this.$(` does not use `this.$el` in options classes but `this.$target` as the source to search into (a bad decision that was taken a very long time ago) https://github.com/odoo/odoo/blob/ed530aef3c6d99712db055099b5f06a31e91d9cf/addons/web_editor/static/src/js/editor/snippets.options.js#L3297

> So it never finds it. I tried to find a case in Website where it does and actually never got past the "if not then return"

Seems to work on my side for the cases I tried, which case are you trying? :thinking:

Comment (Zinston) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3551]: > I think you misses the fact that `this.$(` does not use `this.$el` in options classes but `this.$target` as the source to search into (a bad decision that was taken a very long time ago)

Aaaah thank you, that clarifies it!

> Seems to work on my side for the cases I tried, which case are you trying?

I had put a debugger in `updateUIVisibility` and tried clicking in various snippets, thinking I'd trigger it for the border option which uses `applyTo` but I guess I didn't try in the right places then ;)

Thanks for your help.
